route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (C) 2019-2020 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <ros/ros.h>
+#include <string>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include "route_following_plugin.h"
+
+
+namespace route_following_plugin
+{
+    RouteFollowingPlugin::RouteFollowingPlugin() : mvr_duration_(16.0), current_speed_(0.0) { }
+
+    void RouteFollowingPlugin::initialize()
+    {
+        
+        nh_.reset(new ros::CARMANodeHandle());
+        pnh_.reset(new ros::CARMANodeHandle("~"));
+        
+        plan_maneuver_srv_ = nh_->advertiseService("plugins/RouteFollowingPlugin/plan_maneuvers", &RouteFollowingPlugin::plan_maneuver_cb, this);
+                
+        plugin_discovery_pub_ = nh_->advertise<cav_msgs::Plugin>("plugin_discovery", 1);
+        plugin_discovery_msg_.name = "RouteFollowingPlugin";
+        plugin_discovery_msg_.versionId = "v1.0";
+        plugin_discovery_msg_.available = true;
+        plugin_discovery_msg_.activated = false;
+        plugin_discovery_msg_.type = cav_msgs::Plugin::STRATEGIC;
+        plugin_discovery_msg_.capability = "strategic_plan/plan_maneuvers";
+
+        pose_sub_ = nh_->subscribe("current_pose", 1, &RouteFollowingPlugin::pose_cb, this);
+        twist_sub_ = nh_->subscribe("current_velocity", 1, &RouteFollowingPlugin::twist_cd, this);
+        
+        pnh_->param<double>("minimal_maneuver_duration", mvr_duration_, 16.0);
+
+        wml_.reset(new carma_wm::WMListener());
+        // set world model point form wm listener
+        wm_ = wml_->getWorldModel();
+
+        ros::CARMANodeHandle::setSpinCallback([this]()
+        {
+            plugin_discovery_pub_.publish(plugin_discovery_msg_);
```

Comment:
Is publishing this discovery message with `activated = false` every timestep going to cause issues? If the user activates it from the UI, would publishing this message with `activated = false` deactivate it with the plugin manager?

Reply:
I do not think so. Health Manager only updates a plugin's activity based on user selection (service call from UI). This field will be ignored.